### ISSUE TITLE
Comments in string interpolation expressions break nested interpolation in subsequent expressions.

### ISSUE THREAD

**Issue by back2dos** [HaxeFoundation/haxe#9370]
```haxe
class Main {
  static function main() {
    var foo = 123;
    trace('${true /* breakme */} ${' ${foo} '}');// true  ${foo}
    trace('${true} ${' ${foo} '}');// true  123
  }
}
```

Checked with 4.0.5. and current nightly (4450dfb).

**Comment by Aurel300** [comment]
One more for https://github.com/HaxeFoundation/haxe/issues/2244 ?

**Comment by back2dos** [comment]
Quite possible. I've removed all spaces to make sure they're not the cause: Also, I've printed the expression in a macro and I get this:

```haxe
class Main {
  #if !macro
  static function main() {
    var foo = 123;
    log('${true/*breakme*/}${'${foo}'}');// CString(${true}${'${foo}'},SingleQuotes) -> true${foo}
    log('${true}${'${foo}'}');// CString(${true}${'${foo}'},SingleQuotes) -> true123
  }
  #end
  static macro function log(e) {
    trace(e);
    return macro trace($e);
  }
}
```

It would seem like the lexer throws out the comment right there, and then presumably the position of `${foo}` within the literal no longer lines up with where it was in the file. If interpolation is still governed by this position cache that Simn explained to me a while back, then this misalignment might be what's throwing it off.

**Comment by Simn** [comment]
Yes that's still a thing. We were working towards getting rid of it, but I don't remember what the outcome of that was. Might have been one of these nadako-has-a-branch-things.

**Comment by nadako** [comment]
I have to re-read all the stuff about this. I remember implementing a mini-ast for the interpolated strings, but I did it on the lexer level, which was wrong because we only want to do the interpolation when actually typing the string, not on the token level.

**Comment by RealyUniqueName** [comment]
Can we close this as a duplicate of #2244?

**Comment by back2dos** [comment]
From what I've observed, whitespace is conserved in interpolation expressions, while comments are not. So the cause may be different, although the solution will probably be the same (getting rid of the position cache voodoo).

**Comment by Simn** [comment]
Wasn't the idea to handle this at lexer level (we have to handle it there _somehow_ anyway) and then carry both the complete string and that mini-AST to the typer?

**Comment by nadako** [comment]
See #6475

**Comment by back2dos** [comment]
> I remember implementing a mini-ast for the interpolated strings, but I did it on the lexer level, which was wrong because we only want to do the interpolation when actually typing the string, not on the token level.

There was a pull request for this: https://github.com/HaxeFoundation/haxe/pull/6475

My objection was that the original string would be lost in the process and when you want to process it in a macro, then that gets very awkward.

A (backward compatible) solution would be this:

1. Bring back the proper parser for string interpolation, that also extracts the string as a whole.
2. still just pass it as `CString(theWholeString, SingleQuotes)`
3. add `MacroStringTools.getMiniAst` (or whatever the name) that can produce the mini-AST from *any* string.

Yes, that means that macros that do want the mini-AST have the same string parsed twice, but I strongly suspect that's negligible compared to the cost of encoding, meaning if you do want the mini-AST, the overhead is minimal, and if you don't, then you save the cost of encoding. Also, it would be nice to be able to get the mini-AST for any string (be it a substring of a literal or even something extracted from an external file).

**Comment by nadako** [comment]
Both of these return `true  123` for me now. Could be fixed by #9410? Please check on your side.

**Comment by nadako** [comment]
Yep, seems like it's fixed, nice. And yeah, some macro API for getting the mini-ast would be nice indeed...
